src/firewall/functions.py

```diff
@@ -36,6 +36,24 @@
     i: None for i in range(0, 160) if not (i > 31 and i < 127)
 }
 
+
+###############################################################################
+
+
+class BugError(Exception):
+    """Indicates that there is a bug in the code. You probably don't
+    want to catch this, but fix the bug."""
+
+    @staticmethod
+    def fail(msg=None):
+        if msg is None:
+            msg = "should not be reached"
+        raise BugError("BUG: {msg}")
```

Comment:
`BugError.fail()` is really the `assert_not_reached()` and it looks slightly nicer if this is a function call (instead of a raise)). It's also a bit shorter to type.

Note that it also has a default "msg". In most cases, it's too cumbersome to bother with `if (refcnt <= 0) assert_not_reached("ref-count is non-positive")`. We figure out details about an assertion failure by looking at the backtrace (and line number). The "fail()" method conveys that a bit nicer.

But no strong reasons at all. It could also be `raise BugError()`. I just think it's slightly nicer this way. Look at the usage.